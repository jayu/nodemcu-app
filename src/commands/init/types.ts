export enum AnswerNames {
  setupType = 'setupType',
  projectDir = 'projectDir',
  projectsSubDir = 'projectsSubDir',
  uploadTool = 'uploadTool',
  useCrossCompiler = 'useCrossCompiler',
  crossCompilerPath = 'crossCompilerPath',
  copyLuaCross = 'copyLuaCross',
  commonModulesDir = 'commonModulesDir',
  useLuaRocks = 'useLuaRocks',
  luaRocksModulesDir = 'luaRocksModulesDir'
}

export enum SetupType {
  single = 'single',
  multiple = 'multiple'
}

export enum UploadTool {
  nodemcuUploader = 'nodemcu-uploader',
  nodemcuTool = 'nodemcu-tool',
  decideLater = ''
}

export enum YesOrLater {
  yes = 'yes',
  decideLater = 'decideLater'
}

export type Answers = {
  [AnswerNames.projectDir]: string
  [AnswerNames.setupType]: SetupType
  [AnswerNames.projectsSubDir]?: string
  [AnswerNames.uploadTool]: UploadTool
  [AnswerNames.useCrossCompiler]: YesOrLater
  [AnswerNames.crossCompilerPath]?: string
  [AnswerNames.copyLuaCross]?: boolean
  [AnswerNames.commonModulesDir]?: string
  [AnswerNames.useLuaRocks]: YesOrLater
  [AnswerNames.luaRocksModulesDir]?: string
}

export type SettingsSet = {
  entryDir: string,
  moduleDirs: string[],
  crossCompilerPath?: string,
  uploadToolBinary: UploadTool
}

export type SettingsFile = {
  setupType: SetupType,
  manifestVersion: string,
  default: SettingsSet,
  // Those two are not supported yet
  esp32?: SettingsSet
  esp8266?: SettingsSet
}