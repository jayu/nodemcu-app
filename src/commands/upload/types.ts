export type UploaderWrapperDataType = {
  port: string,
  baudRate: number,
  files: string[][],
  cwd: string
}

export type UploaderWrapperFNType = (data: UploaderWrapperDataType) => Promise<string | null>