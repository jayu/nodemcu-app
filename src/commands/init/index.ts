import commander from 'commander'
import init from './init'

function create(program: commander.Command) {
  program
    .command('init')
    .description(
      'Run interactive wizard that will create nodemcu-app project structure.'
    )
    .action(async () => {
      await init()
    })
}

export default create
