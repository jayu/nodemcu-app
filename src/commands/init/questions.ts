import { PromptObject } from 'prompts'
import fs from 'fs'
import { getAbsolutePath } from '../../utils'

import {AnswerNames, SetupType, Answers, YesOrLater, UploadTool } from './types';

const nonEmptyValidator = (value: string) => {
  return value.trim().length > 0 ? true : 'Value cannot be empty.'
}

const questions = [
  {
    type: 'select',
    name: AnswerNames.setupType,
    message: 'Select setup type',
    choices: [
      {
        title: 'Multiple',
        description:
          'Create a directory to develop multiple projects with shared modules (recommended)',
        value: SetupType.multiple
      },
      {
        title: 'Single',
        description: 'Create single project directory without shared modules',
        value: SetupType.single
      }
    ],
    initial: 0
  },
  {
    type: 'text',
    name: AnswerNames.projectDir,
    message: (setupType: SetupType) =>
      setupType === SetupType.multiple
        ? `Please provide a name for a directory`
        : 'Please provide project name',
    initial: 'my-nodemcu-app',
    validate: nonEmptyValidator
  },
  {
    type: (_prev: string, answers: Pick<Answers, AnswerNames.setupType>) =>
      answers[AnswerNames.setupType] === SetupType.multiple ? 'text' : null,
    name: AnswerNames.projectsSubDir,
    message: 'Please provide a name of a subdirectory for projects',
    initial: 'projects',
    validate: nonEmptyValidator
  },
  {
    type: (_prev: string, answers: Pick<Answers, AnswerNames.setupType>) =>
      answers[AnswerNames.setupType] === SetupType.multiple ? 'text' : null,
    name: AnswerNames.commonModulesDir,
    message: 'Please provide a name for a directory to store common modules',
    initial: 'modules',
    validate: nonEmptyValidator
  },
  {
    type: 'select',
    name: AnswerNames.useLuaRocks,
    message: 'Would you like to use LuaRocks (Lua package manager) modules?',
    choices: [
      {
        title: 'yes',
        description: 'The `luarocks` binary must be available in system PATH',
        value: YesOrLater.yes
      },
      {
        title: 'I will decide later',
        description:
          'You can set proper value in generated setup file at any time',
        value: YesOrLater.decideLater
      }
    ],
    initial: 0
  },
  {
    type: (prev: YesOrLater) => (prev === YesOrLater.yes ? 'text' : null),
    name: AnswerNames.luaRocksModulesDir,
    message: 'Please provide a name for a directory to store LuaRocks modules',
    initial: 'luarocks_modules',
    validate: nonEmptyValidator
  },
  {
    type: 'select',
    name: AnswerNames.uploadTool,
    message:
      'Select a tool for uploading code on nodeMCU (tool binary must be in system PATH)',
    choices: [
      {
        title: 'nodemcu-uploader',
        description:
          'Home page: https://github.com/kmpm/nodemcu-uploader. Requires Python. Fastest upload.',
        value: UploadTool.nodemcuUploader
      },
      {
        title: 'I wll decide later',
        description: 'You can tool name in generated setup file at any time',
        value: UploadTool.decideLater
      },
      {
        title: 'nodemcu-tool (integration available soon)',
        description:
          'Home page: https://github.com/AndiDittrich/NodeMCU-Tool. Requires JS. Quite slow upload.',
        value: UploadTool.nodemcuTool,
        disabled: true
      }
    ],
    initial: 0
  },
  {
    type: 'select',
    name: AnswerNames.useCrossCompiler,
    message: 'Would you like to use compiler or LFS (Lua File Store) features?',
    choices: [
      {
        title: 'yes',
        description:
          'You will need to build luac.cross file on your system (https://nodemcu.readthedocs.io/en/dev/getting-started/#build-luaccross)',
        value: YesOrLater.yes
      },
      {
        title: 'I will decide later',
        description:
          'You can set proper value in generated setup file at any time',
        value: YesOrLater.decideLater
      }
    ],
    initial: 0
  },
  {
    type: (prev: YesOrLater) => (prev === YesOrLater.yes ? 'text' : null),
    name: AnswerNames.crossCompilerPath,
    message: 'Provide path to luac.cross binary file.',
    format: (value: string) => {
      return getAbsolutePath(value)
    },
    validate: (value: string) => {
      return fs.existsSync(getAbsolutePath(value))
        ? true
        : 'Path do not exist! Please try again.'
    }
    // initial: './luac.cross', // default value is buggy when user provide incorrect input
  },
  {
    type: (
      _prev: string,
      answers: Pick<Answers, AnswerNames.useCrossCompiler>
    ) =>
      answers[AnswerNames.useCrossCompiler] === YesOrLater.yes
        ? 'confirm'
        : null,
    name: AnswerNames.copyLuaCross,
    message:
      'Would you like to copy luac.cross into project directory (recommended)?',
    initial: true
  }
]


export default questions as PromptObject[]
