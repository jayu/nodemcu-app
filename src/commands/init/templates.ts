import dedent from 'dedent'

export const exampleModule = dedent`
local function increment(value)
  return value + 1
end

local function decrement(value)
  return value - 1
end

return {
  increment = increment, 
  decrement = decrement
}

`
export const exampleModuleName = 'exampleModule'

export const exampleInitFile = dedent`
local exampleModule = require('${exampleModuleName}')

print('Welcome in example nodemcu-app project')

local counter = 0
print('Counter value: ' .. counter)
counter = exampleModule.increment(counter)
print('Counter value incremented by example module: ' .. counter)

`

export const gitignore = dedent`
dist
`
