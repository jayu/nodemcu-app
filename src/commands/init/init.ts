import prompts from 'prompts'
import path from 'path'
import fs from 'fs'
import rimraf from 'rimraf'

import questions from './questions'

import {
  Answers,
  YesOrLater,
  SetupType,
  SettingsFile
} from './types';

import * as templates from './templates'

export const createProject = (response: Answers) => {
  const tempDirPath = fs.mkdtempSync('nodemcu-app')
  try {
    let crossCompilerPath = ''

    if (
      response.useCrossCompiler === YesOrLater.yes &&
      response.crossCompilerPath !== undefined
    ) {
      if (response.copyLuaCross) {
        crossCompilerPath = './luac.cross'
        fs.copyFileSync(
          response.crossCompilerPath,
          path.join(tempDirPath, crossCompilerPath)
        )
      } else {
        crossCompilerPath = response.crossCompilerPath
      }
    }
    const moduleDirs: string[] = []

    if (response.setupType === SetupType.multiple && response.commonModulesDir) {
      moduleDirs.push(`./${response.commonModulesDir}`)
    }

    if (response.useLuaRocks === YesOrLater.yes && response.luaRocksModulesDir) {
      moduleDirs.push(`./${response.luaRocksModulesDir}`)
    }

    const entryDir =
      response.setupType === SetupType.multiple
        ? `./${response.projectsSubDir}`
        : './src'

    const settingsJSON: SettingsFile = {
      setupType: response.setupType,
      manifestVersion: '1.0',
      default: {
        entryDir,
        moduleDirs,
        crossCompilerPath,
        uploadToolBinary: response.uploadTool
      }
    }

    const dirsToCreate = [...moduleDirs, entryDir]
    // Create all sub-directories
    dirsToCreate.forEach((dir) => {
      fs.mkdirSync(path.join(tempDirPath, dir))
    })

    // Save settings.json file
    const settingsJSONString = JSON.stringify(settingsJSON, undefined, 2)
    fs.writeFileSync(
      path.join(tempDirPath, 'settings.json'),
      settingsJSONString
    )

    // Create an example module
    fs.writeFileSync(
      path.join(
        tempDirPath,
        response.setupType === SetupType.multiple && response.commonModulesDir
          ? response.commonModulesDir
          : entryDir,
        `${templates.exampleModuleName}.lua`
      ),
      templates.exampleModule
    )

    // Create initial project directory for multiple setup
    const exampleProjectDirectory =
      response.setupType === SetupType.multiple ? 'example-project' : ''

    if (exampleProjectDirectory !== '') {
      fs.mkdirSync(path.join(tempDirPath, entryDir, exampleProjectDirectory))
    }

    // Create example project entry file
    fs.writeFileSync(
      path.join(tempDirPath, entryDir, exampleProjectDirectory, 'init.lua'),
      templates.exampleInitFile
    )

    // Create gitignore for dist directories
    fs.writeFileSync(
      path.join(tempDirPath, '.gitignore'),
      templates.gitignore
    )

    // Move temp directory to target project directory
    fs.renameSync(tempDirPath, response.projectDir)
    return {
      success: true,
      dirName: `./${response.projectDir}`,
      settingsString: settingsJSONString
    }
  } catch (error) {
    rimraf.sync(tempDirPath)
    return {
      success: false,
      error
    }
  }
}

async function init() {
  let promptsCancelled = false
  const onCancel = () => {
    promptsCancelled = true
  }
  const response: Answers = (await prompts(questions, { onCancel })) as any
  if (!promptsCancelled) {
    const { success, dirName, settingsString, error } = createProject(response)
    if (success) {
      console.log(`Project created in ${dirName}`)
      console.log('With settings:\n', settingsString)
    }
    else {
      console.error('An error occurred during project setup:', error)
    }
  }
  else {
    console.log('Cancelled project creation')
  }
}

export default init
