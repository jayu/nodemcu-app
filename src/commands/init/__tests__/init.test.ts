import fs from 'fs';
import mock from 'mock-fs';
import { createProject } from '../init';
import { SetupType, UploadTool, YesOrLater, AnswerNames } from '../types'
const defaultProjectName = 'test-project';

const defaultAnswers = {
  [AnswerNames.projectDir]: defaultProjectName,
  [AnswerNames.setupType]: SetupType.multiple,
  [AnswerNames.projectsSubDir]: 'src',
  [AnswerNames.uploadTool]: UploadTool.nodemcuUploader,
  [AnswerNames.useCrossCompiler]: YesOrLater.decideLater,
  [AnswerNames.commonModulesDir]: 'modules',
  [AnswerNames.useLuaRocks]: false,
}

const getSettingsJSON = (projectName: string) => {
  return JSON.parse(fs.readFileSync(`${projectName}/settings.json`).toString())
}

const readDir = (dir: string) => fs.readdirSync(dir, { withFileTypes: true })

type Tree = Array<{ name: string, isDirectory: boolean, children: Tree | null }>

const getFilesTree = (root: string): Tree => {
  const dir = readDir(root);
  return dir.reduce<Tree>((tree, entry) => {
    return [
      ...tree,
      {
        name: entry.name,
        isDirectory: entry.isDirectory(),
        children: entry.isDirectory() ? getFilesTree(`${root}/${entry.name}`) : null
      }
    ]
  }, [])
}

beforeEach(() => {
  mock({
    '/': {}
  })
})

afterEach(() => {
  mock.restore()
})

it('Should create example project with setup type multiple', () => {
  createProject(defaultAnswers);
  const settings = getSettingsJSON(defaultProjectName)
  const filesTree = getFilesTree(defaultProjectName)
  mock.restore()
  expect(settings.setupType).toBe('multiple')
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})

it('Should create example project with setup type single', () => {
  createProject({ ...defaultAnswers, [AnswerNames.setupType]: SetupType.single });
  const settings = getSettingsJSON(defaultProjectName)
  const filesTree = getFilesTree(defaultProjectName)
  mock.restore()
  expect(settings.setupType).toBe('single')
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})

it('Should create example project with setup type multiple and custom dir names', () => {
  const projectName = 'my-project'
  createProject({
    ...defaultAnswers,
    [AnswerNames.projectDir]: 'my-project',
    [AnswerNames.projectsSubDir]: 'source-code',
    [AnswerNames.commonModulesDir]: 'custom-modules',
  });
  const settings = getSettingsJSON(projectName)
  const filesTree = getFilesTree(projectName)
  mock.restore()
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})

it('Should create example project with cross compiler without copying it', () => {
  createProject({
    ...defaultAnswers,
    [AnswerNames.useCrossCompiler]: YesOrLater.yes,
    [AnswerNames.crossCompilerPath]: '/some-dir/luac.cross',
    [AnswerNames.copyLuaCross]: false,
  });
  const settings = getSettingsJSON(defaultProjectName)
  const filesTree = getFilesTree(defaultProjectName)
  mock.restore()
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})

it('Should create example project with cross compiler with copying it', () => {

  mock({
    '/some-dir/my-luac.cross': 'fake-binary-content'
  })
  createProject({
    ...defaultAnswers,
    [AnswerNames.useCrossCompiler]: YesOrLater.yes,
    [AnswerNames.crossCompilerPath]: '/some-dir/my-luac.cross',
    [AnswerNames.copyLuaCross]: true,
  });
  const settings = getSettingsJSON(defaultProjectName)
  const filesTree = getFilesTree(defaultProjectName)
  mock.restore()
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})

it('Should create example project with using of luarocks modules', () => {

  mock({
    '/some-dir/my-luac.cross': 'fake-binary-content'
  })
  createProject({
    ...defaultAnswers,
    [AnswerNames.useLuaRocks]: true,
    [AnswerNames.luaRocksModulesDir]: 'luarocks-modules-dir'
  });
  const settings = getSettingsJSON(defaultProjectName)
  const filesTree = getFilesTree(defaultProjectName)
  mock.restore()
  expect(settings).toMatchSnapshot()
  expect(filesTree).toMatchSnapshot()
})