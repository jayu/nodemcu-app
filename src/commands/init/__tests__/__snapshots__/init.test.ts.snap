// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should create example project with cross compiler with copying it 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "./luac.cross",
    "entryDir": "./src",
    "moduleDirs": Array [
      "./modules",
    ],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with cross compiler with copying it 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "luac.cross",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create example project with cross compiler without copying it 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "/some-dir/luac.cross",
    "entryDir": "./src",
    "moduleDirs": Array [
      "./modules",
    ],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with cross compiler without copying it 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create example project with setup type multiple 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./src",
    "moduleDirs": Array [
      "./modules",
    ],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with setup type multiple 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create example project with setup type multiple and custom dir names 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./source-code",
    "moduleDirs": Array [
      "./custom-modules",
    ],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with setup type multiple and custom dir names 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "custom-modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "source-code",
  },
]
`;

exports[`Should create example project with setup type multiple nodemcu-tool uploader 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./src",
    "moduleDirs": Array [
      "./modules",
    ],
    "uploadToolBinary": "nodemcu-tool",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with setup type multiple nodemcu-tool uploader 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create example project with setup type single 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./src",
    "moduleDirs": Array [],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "single",
}
`;

exports[`Should create example project with setup type single 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
      Object {
        "children": null,
        "isDirectory": false,
        "name": "init.lua",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create example project with using of luarocks modules 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./src",
    "moduleDirs": Array [
      "./modules",
      "./luarocks-modules-dir",
    ],
    "uploadToolBinary": "nodemcu-uploader",
  },
  "manifestVersion": "1.1",
  "setupType": "multiple",
}
`;

exports[`Should create example project with using of luarocks modules 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": Array [],
    "isDirectory": true,
    "name": "luarocks-modules-dir",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
    ],
    "isDirectory": true,
    "name": "modules",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": null,
            "isDirectory": false,
            "name": "init.lua",
          },
        ],
        "isDirectory": true,
        "name": "example-project",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;

exports[`Should create project with setup type single,  nodemcu-tool, and no other options 1`] = `
Object {
  "default": Object {
    "crossCompilerPath": "",
    "entryDir": "./src",
    "moduleDirs": Array [],
    "uploadToolBinary": "nodemcu-tool",
  },
  "manifestVersion": "1.1",
  "setupType": "single",
}
`;

exports[`Should create project with setup type single,  nodemcu-tool, and no other options 2`] = `
Array [
  Object {
    "children": null,
    "isDirectory": false,
    "name": ".gitignore",
  },
  Object {
    "children": null,
    "isDirectory": false,
    "name": "settings.json",
  },
  Object {
    "children": Array [
      Object {
        "children": null,
        "isDirectory": false,
        "name": "exampleModule.lua",
      },
      Object {
        "children": null,
        "isDirectory": false,
        "name": "init.lua",
      },
    ],
    "isDirectory": true,
    "name": "src",
  },
]
`;
